client2/send.go

```diff
@@ -0,0 +1,162 @@
+				return nil, nil, 0, err
+			}
+		}
+
+		// If the path selection process ends up straddling an epoch
+		// transition, then redo the path selection.
+		if time.Since(start) > budget {
+			continue
+		}
+
+		// It is possible, but unlikely that a series of delays exceeding
+		// the PKI publication imposted limitations will be selected.  When
+		// that happens, the path selection must be redone.
+		if then.Sub(now) < epochtime.Period*2 {
+			if request.WithSURB {
+				payload := make([]byte, 2, 2+c.geo.SURBLength+len(request.Payload))
+				payload[0] = 1 // Packet has a SURB.
+				surb, k, err := c.sphinx.NewSURB(rand.Reader, revPath)
+				if err != nil {
+					return nil, nil, 0, err
+				}
+				payload = append(payload, surb...)
+				payload = append(payload, request.Payload...)
+
+				blob := make([]byte, c.geo.ForwardPayloadLength)
+				copy(blob, payload)
+
+				pkt, err := c.sphinx.NewPacket(rand.Reader, fwdPath, blob)
+				if err != nil {
+					return nil, nil, 0, err
+				}
+				return pkt, k, then.Sub(now), err
+			} else {
+				blob := make([]byte, c.geo.ForwardPayloadLength)
+				copy(blob, payload)
+
+				pkt, err := c.sphinx.NewPacket(rand.Reader, fwdPath, blob)
+				if err != nil {
+					return nil, nil, 0, err
+				}
+				return pkt, nil, then.Sub(now), nil
+			}
+		}
+	}
+}
+
+// SendCiphertext sends the ciphertext b to the recipient/provider, with a
+// SURB identified by surbID, and returns the SURB decryption key and total
+// round trip delay.
+func (c *Client) SendCiphertext(request *Request) ([]byte, time.Duration, error) {
+	pkt, k, rtt, err := c.ComposeSphinxPacket(request)
+	if err != nil {
+		panic(fmt.Sprintf("COMPOSE SPHINX PACKET FAIL %s", err.Error()))
+	}
+	err = c.conn.sendPacket(pkt)
+	return k, rtt, err
+}
+
+func (c *Client) makePath(recipient []byte, provider *[32]byte, surbID *[sConstants.SURBIDLength]byte, baseTime time.Time, isForward bool) ([]*path.PathHop, time.Time, error) {
+	if c.conn.provider == nil {
+		panic("source provider cannot be nil")
+	}
+	srcProvider, dstProvider := c.conn.provider, provider
+	if !isForward {
+		srcProvider, dstProvider = dstProvider, srcProvider
+	}
+
+	// Get the current PKI document.
+	doc := c.CurrentDocument()
+	if doc == nil {
+		return nil, time.Time{}, newPKIError("client2: no PKI document for current epoch")
+	}
+
+	// Get the descriptors.
+	src, err := doc.GetProviderByKeyHash(srcProvider)
+	if err != nil {
+		return nil, time.Time{}, newPKIError("client2: failed to find source Provider: %v", err)
+	}
+	dst, err := doc.GetProviderByKeyHash(dstProvider)
+	if err != nil {
+		return nil, time.Time{}, newPKIError("client2: failed to find destination Provider: %v", err)
+	}
+
+	rng := rand.NewMath()
+	p, t, err := path.New(rng, c.cfg.SphinxGeometry, doc, recipient, src, dst, surbID, baseTime, true, isForward)
+	if err == nil {
+		c.logPath(doc, p)
+	}
+
+	return p, t, err
+}
+
+func (c *Client) logPath(doc *cpki.Document, p []*path.PathHop) error {
```

Comment:
instead duplicating this from `decoy` we should probably have it be a method of `path`, taking a logger as argument?